solr/core/src/java/org/apache/solr/search/CaffeineCache.java

```diff
@@ -176,33 +193,92 @@ public V get(K key) {
     return cache.getIfPresent(key);
   }
 
-  @Override
-  public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
-    return cache.get(key, k -> {
-      inserts.increment();
-      V value = mappingFunction.apply(k);
-      if (value == null) {
-        return null;
+  private V computeAsync(K key, IOFunction<? super K, ? extends V> mappingFunction) throws IOException {
+    CompletableFuture<V> future = new CompletableFuture<>();
+    CompletableFuture<V> result = asyncCache.asMap().putIfAbsent(key, future);
+    if (result != null) {
+      try {
+        // Another thread is already working on this computation, wait for them to finish
+        return result.join();
```

Comment:
@ben-manes this code path doesn't seem to record a hit in the cache stats for the "happy" case where the result is already completed normally. should I be using a different API?

Reply:
I think stats on an async cache, especially with the `asMap()` view, is confusing. I don't know if I got it instrumented right. It tries to be good enough but either requires a deeper review to fix holes, or custom stats to match your desired reporting.